{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: appdirs==1.4.4 in /home/codespace/.python/current/lib/python3.12/site-packages (from -r requirements.txt (line 1)) (1.4.4)\n",
      "Requirement already satisfied: colorama==0.4.6 in /home/codespace/.local/lib/python3.12/site-packages (from -r requirements.txt (line 2)) (0.4.6)\n",
      "Requirement already satisfied: dash==2.0.0 in /home/codespace/.python/current/lib/python3.12/site-packages (from -r requirements.txt (line 3)) (2.0.0)\n",
      "Requirement already satisfied: dash-core-components==2.0.0 in /home/codespace/.python/current/lib/python3.12/site-packages (from -r requirements.txt (line 4)) (2.0.0)\n",
      "Requirement already satisfied: dash-html-components==2.0.0 in /home/codespace/.python/current/lib/python3.12/site-packages (from -r requirements.txt (line 5)) (2.0.0)\n",
      "Requirement already satisfied: dash-table==5.0.0 in /home/codespace/.python/current/lib/python3.12/site-packages (from -r requirements.txt (line 6)) (5.0.0)\n",
      "Requirement already satisfied: Flask==2.1.1 in /home/codespace/.python/current/lib/python3.12/site-packages (from -r requirements.txt (line 7)) (2.1.1)\n",
      "Requirement already satisfied: Flask-Compress==1.5.0 in /home/codespace/.python/current/lib/python3.12/site-packages (from -r requirements.txt (line 8)) (1.5.0)\n",
      "Requirement already satisfied: gunicorn==20.0.4 in /home/codespace/.python/current/lib/python3.12/site-packages (from -r requirements.txt (line 9)) (20.0.4)\n",
      "Requirement already satisfied: Jinja2==3.1.2 in /home/codespace/.python/current/lib/python3.12/site-packages (from -r requirements.txt (line 10)) (3.1.2)\n",
      "Requirement already satisfied: lxml==4.9.3 in /home/codespace/.python/current/lib/python3.12/site-packages (from -r requirements.txt (line 11)) (4.9.3)\n",
      "Requirement already satisfied: numpy==1.26.4 in /home/codespace/.python/current/lib/python3.12/site-packages (from -r requirements.txt (line 12)) (1.26.4)\n",
      "Requirement already satisfied: pandas==2.2.2 in /home/codespace/.python/current/lib/python3.12/site-packages (from -r requirements.txt (line 13)) (2.2.2)\n",
      "Requirement already satisfied: plotly==5.4.0 in /home/codespace/.python/current/lib/python3.12/site-packages (from -r requirements.txt (line 14)) (5.4.0)\n",
      "Requirement already satisfied: python-dotenv==1.0.0 in /home/codespace/.python/current/lib/python3.12/site-packages (from -r requirements.txt (line 15)) (1.0.0)\n",
      "Requirement already satisfied: scikit-learn==1.4.1.post1 in /home/codespace/.python/current/lib/python3.12/site-packages (from -r requirements.txt (line 16)) (1.4.1.post1)\n",
      "Requirement already satisfied: scipy==1.13.0 in /home/codespace/.python/current/lib/python3.12/site-packages (from -r requirements.txt (line 17)) (1.13.0)\n",
      "Requirement already satisfied: urllib3==2.0.4 in /home/codespace/.python/current/lib/python3.12/site-packages (from -r requirements.txt (line 18)) (2.0.4)\n",
      "Requirement already satisfied: Werkzeug==2.0.3 in /home/codespace/.python/current/lib/python3.12/site-packages (from -r requirements.txt (line 19)) (2.0.3)\n",
      "Collecting yfinance==0.2.55 (from -r requirements.txt (line 20))\n",
      "  Downloading yfinance-0.2.55-py2.py3-none-any.whl.metadata (5.8 kB)\n",
      "Requirement already satisfied: itsdangerous>=2.0 in /home/codespace/.python/current/lib/python3.12/site-packages (from Flask==2.1.1->-r requirements.txt (line 7)) (2.2.0)\n",
      "Requirement already satisfied: click>=8.0 in /home/codespace/.python/current/lib/python3.12/site-packages (from Flask==2.1.1->-r requirements.txt (line 7)) (8.1.8)\n",
      "Requirement already satisfied: brotli in /home/codespace/.python/current/lib/python3.12/site-packages (from Flask-Compress==1.5.0->-r requirements.txt (line 8)) (1.1.0)\n",
      "Requirement already satisfied: setuptools>=3.0 in /home/codespace/.local/lib/python3.12/site-packages (from gunicorn==20.0.4->-r requirements.txt (line 9)) (76.0.0)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in /home/codespace/.local/lib/python3.12/site-packages (from Jinja2==3.1.2->-r requirements.txt (line 10)) (3.0.2)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in /home/codespace/.local/lib/python3.12/site-packages (from pandas==2.2.2->-r requirements.txt (line 13)) (2.9.0.post0)\n",
      "Requirement already satisfied: pytz>=2020.1 in /home/codespace/.local/lib/python3.12/site-packages (from pandas==2.2.2->-r requirements.txt (line 13)) (2025.1)\n",
      "Requirement already satisfied: tzdata>=2022.7 in /home/codespace/.local/lib/python3.12/site-packages (from pandas==2.2.2->-r requirements.txt (line 13)) (2025.1)\n",
      "Requirement already satisfied: tenacity>=6.2.0 in /home/codespace/.python/current/lib/python3.12/site-packages (from plotly==5.4.0->-r requirements.txt (line 14)) (9.1.2)\n",
      "Requirement already satisfied: six in /home/codespace/.local/lib/python3.12/site-packages (from plotly==5.4.0->-r requirements.txt (line 14)) (1.17.0)\n",
      "Requirement already satisfied: joblib>=1.2.0 in /home/codespace/.local/lib/python3.12/site-packages (from scikit-learn==1.4.1.post1->-r requirements.txt (line 16)) (1.4.2)\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in /home/codespace/.local/lib/python3.12/site-packages (from scikit-learn==1.4.1.post1->-r requirements.txt (line 16)) (3.6.0)\n",
      "Requirement already satisfied: requests>=2.31 in /home/codespace/.local/lib/python3.12/site-packages (from yfinance==0.2.55->-r requirements.txt (line 20)) (2.32.3)\n",
      "Requirement already satisfied: multitasking>=0.0.7 in /home/codespace/.python/current/lib/python3.12/site-packages (from yfinance==0.2.55->-r requirements.txt (line 20)) (0.0.11)\n",
      "Requirement already satisfied: platformdirs>=2.0.0 in /home/codespace/.local/lib/python3.12/site-packages (from yfinance==0.2.55->-r requirements.txt (line 20)) (4.3.6)\n",
      "Requirement already satisfied: frozendict>=2.3.4 in /home/codespace/.python/current/lib/python3.12/site-packages (from yfinance==0.2.55->-r requirements.txt (line 20)) (2.4.6)\n",
      "Collecting peewee>=3.16.2 (from yfinance==0.2.55->-r requirements.txt (line 20))\n",
      "  Downloading peewee-3.17.9.tar.gz (3.0 MB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.0/3.0 MB\u001b[0m \u001b[31m36.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25h  Installing build dependencies ... \u001b[?25ldone\n",
      "\u001b[?25h  Getting requirements to build wheel ... \u001b[?25ldone\n",
      "\u001b[?25h  Preparing metadata (pyproject.toml) ... \u001b[?25ldone\n",
      "\u001b[?25hRequirement already satisfied: beautifulsoup4>=4.11.1 in /home/codespace/.local/lib/python3.12/site-packages (from yfinance==0.2.55->-r requirements.txt (line 20)) (4.13.3)\n",
      "Requirement already satisfied: soupsieve>1.2 in /home/codespace/.local/lib/python3.12/site-packages (from beautifulsoup4>=4.11.1->yfinance==0.2.55->-r requirements.txt (line 20)) (2.6)\n",
      "Requirement already satisfied: typing-extensions>=4.0.0 in /home/codespace/.local/lib/python3.12/site-packages (from beautifulsoup4>=4.11.1->yfinance==0.2.55->-r requirements.txt (line 20)) (4.12.2)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /home/codespace/.local/lib/python3.12/site-packages (from requests>=2.31->yfinance==0.2.55->-r requirements.txt (line 20)) (3.4.1)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /home/codespace/.local/lib/python3.12/site-packages (from requests>=2.31->yfinance==0.2.55->-r requirements.txt (line 20)) (3.10)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /home/codespace/.local/lib/python3.12/site-packages (from requests>=2.31->yfinance==0.2.55->-r requirements.txt (line 20)) (2025.1.31)\n",
      "Downloading yfinance-0.2.55-py2.py3-none-any.whl (109 kB)\n",
      "Building wheels for collected packages: peewee\n",
      "  Building wheel for peewee (pyproject.toml) ... \u001b[?25ldone\n",
      "\u001b[?25h  Created wheel for peewee: filename=peewee-3.17.9-cp312-cp312-linux_x86_64.whl size=1029257 sha256=6a89c898cea1fdf27c788326e1dacd53ccdeb5c7035e7373d7baa5bcf8294952\n",
      "  Stored in directory: /home/codespace/.cache/pip/wheels/43/ef/2d/2c51d496bf084945ffdf838b4cc8767b8ba1cc20eb41588831\n",
      "Successfully built peewee\n",
      "Installing collected packages: peewee, yfinance\n",
      "  Attempting uninstall: yfinance\n",
      "    Found existing installation: yfinance 0.2.28\n",
      "    Uninstalling yfinance-0.2.28:\n",
      "      Successfully uninstalled yfinance-0.2.28\n",
      "Successfully installed peewee-3.17.9 yfinance-0.2.55\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install -r requirements.txt\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: gevent in /usr/local/python/3.12.1/lib/python3.12/site-packages (24.11.1)\n",
      "Requirement already satisfied: zope.event in /usr/local/python/3.12.1/lib/python3.12/site-packages (from gevent) (5.0)\n",
      "Requirement already satisfied: zope.interface in /usr/local/python/3.12.1/lib/python3.12/site-packages (from gevent) (7.2)\n",
      "Requirement already satisfied: greenlet>=3.1.1 in /usr/local/python/3.12.1/lib/python3.12/site-packages (from gevent) (3.1.1)\n",
      "Requirement already satisfied: setuptools in /home/codespace/.local/lib/python3.12/site-packages (from zope.event->gevent) (76.0.0)\n"
     ]
    }
   ],
   "source": [
    "! pip install gevent"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "::1 - - [2025-04-09 17:38:16] \"GET / HTTP/1.1\" 200 1988 0.278949\n",
      "::1 - - [2025-04-09 17:38:16] \"GET /assets/styles.css?m=1744171046.4764292 HTTP/1.1\" 304 141 0.000857\n",
      "::1 - - [2025-04-09 17:38:16] \"GET /_dash-dependencies HTTP/1.1\" 200 1253 0.000834\n",
      "::1 - - [2025-04-09 17:38:16] \"GET /_dash-layout HTTP/1.1\" 200 2647 0.000893\n",
      "::1 - - [2025-04-09 17:38:16] \"POST /_dash-update-component HTTP/1.1\" 200 387 0.000819\n",
      "::1 - - [2025-04-09 17:38:16] \"POST /_dash-update-component HTTP/1.1\" 200 170 0.001076\n",
      "::1 - - [2025-04-09 17:38:16] \"POST /_dash-update-component HTTP/1.1\" 200 166 0.001101\n",
      "::1 - - [2025-04-09 17:38:17] \"GET /_dash-component-suites/dash/dcc/async-datepicker.js HTTP/1.1\" 304 66 0.000804\n",
      "::1 - - [2025-04-09 17:38:17] \"GET /assets/stocks-img.jpg HTTP/1.1\" 304 145 0.000668\n",
      "::1 - - [2025-04-09 17:38:17] \"POST /_dash-update-component HTTP/1.1\" 200 168 0.000411\n",
      "::1 - - [2025-04-09 17:38:28] \"POST /_dash-update-component HTTP/1.1\" 200 166 0.000558\n",
      "::1 - - [2025-04-09 17:38:28] \"POST /_dash-update-component HTTP/1.1\" 200 168 0.000498\n",
      "::1 - - [2025-04-09 17:38:30] \"POST /_dash-update-component HTTP/1.1\" 200 168 0.000499\n",
      "::1 - - [2025-04-09 17:38:30] \"POST /_dash-update-component HTTP/1.1\" 200 166 0.000356\n",
      "::1 - - [2025-04-09 17:38:34] \"POST /_dash-update-component HTTP/1.1\" 200 1892 0.383770\n",
      "::1 - - [2025-04-09 17:38:34] \"POST /_dash-update-component HTTP/1.1\" 200 170 0.000510\n",
      "::1 - - [2025-04-09 17:38:34] \"POST /_dash-update-component HTTP/1.1\" 200 168 0.000367\n",
      "::1 - - [2025-04-09 17:38:34] \"POST /_dash-update-component HTTP/1.1\" 200 166 0.000360\n",
      "YF.download() has changed argument auto_adjust default to True\n",
      "[*********************100%***********************]  1 of 1 completed\n",
      "[2025-04-09 17:38:39,993] ERROR in app: Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/flask/app.py\", line 2077, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/flask/app.py\", line 1525, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/flask/app.py\", line 1523, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/flask/app.py\", line 1509, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/dash/dash.py\", line 1336, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/dash/_callback.py\", line 151, in add_context\n",
      "    output_value = func(*func_args, **func_kwargs)  # %% callback invoked %%\n",
      "                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/workspaces/CS329E/stock-prediction-dash/app.py\", line 150, in stock_price\n",
      "    fig = get_stock_price_fig(df)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/workspaces/CS329E/stock-prediction-dash/app.py\", line 19, in get_stock_price_fig\n",
      "    fig = px.line(df,\n",
      "          ^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/plotly/express/_chart_types.py\", line 260, in line\n",
      "    return make_figure(args=locals(), constructor=go.Scatter)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/plotly/express/_core.py\", line 1933, in make_figure\n",
      "    args = build_dataframe(args, constructor)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/plotly/express/_core.py\", line 1405, in build_dataframe\n",
      "    df_output, wide_id_vars = process_args_into_dataframe(\n",
      "                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/plotly/express/_core.py\", line 1222, in process_args_into_dataframe\n",
      "    df_output[col_name] = to_unindexed_series(df_input[argument])\n",
      "                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/plotly/express/_core.py\", line 1073, in to_unindexed_series\n",
      "    return pd.Series(x).reset_index(drop=True)\n",
      "           ^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/pandas/core/series.py\", line 584, in __init__\n",
      "    data = sanitize_array(data, index, dtype, copy)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/pandas/core/construction.py\", line 633, in sanitize_array\n",
      "    return sanitize_array(\n",
      "           ^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/pandas/core/construction.py\", line 659, in sanitize_array\n",
      "    subarr = _sanitize_ndim(subarr, data, dtype, index, allow_2d=allow_2d)\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/pandas/core/construction.py\", line 718, in _sanitize_ndim\n",
      "    raise ValueError(\n",
      "ValueError: Data must be 1-dimensional, got ndarray of shape (2462, 1) instead\n",
      "::1 - - [2025-04-09 17:38:39] \"POST /_dash-update-component HTTP/1.1\" 500 426 0.202139\n",
      "[*********************100%***********************]  1 of 1 completed\n",
      "[2025-04-09 17:38:48,906] ERROR in app: Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/flask/app.py\", line 2077, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/flask/app.py\", line 1525, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/flask/app.py\", line 1523, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/flask/app.py\", line 1509, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/dash/dash.py\", line 1336, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/dash/_callback.py\", line 151, in add_context\n",
      "    output_value = func(*func_args, **func_kwargs)  # %% callback invoked %%\n",
      "                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/workspaces/CS329E/stock-prediction-dash/app.py\", line 150, in stock_price\n",
      "    fig = get_stock_price_fig(df)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/workspaces/CS329E/stock-prediction-dash/app.py\", line 19, in get_stock_price_fig\n",
      "    fig = px.line(df,\n",
      "          ^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/plotly/express/_chart_types.py\", line 260, in line\n",
      "    return make_figure(args=locals(), constructor=go.Scatter)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/plotly/express/_core.py\", line 1933, in make_figure\n",
      "    args = build_dataframe(args, constructor)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/plotly/express/_core.py\", line 1405, in build_dataframe\n",
      "    df_output, wide_id_vars = process_args_into_dataframe(\n",
      "                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/plotly/express/_core.py\", line 1222, in process_args_into_dataframe\n",
      "    df_output[col_name] = to_unindexed_series(df_input[argument])\n",
      "                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/plotly/express/_core.py\", line 1073, in to_unindexed_series\n",
      "    return pd.Series(x).reset_index(drop=True)\n",
      "           ^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/pandas/core/series.py\", line 584, in __init__\n",
      "    data = sanitize_array(data, index, dtype, copy)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/pandas/core/construction.py\", line 633, in sanitize_array\n",
      "    return sanitize_array(\n",
      "           ^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/pandas/core/construction.py\", line 659, in sanitize_array\n",
      "    subarr = _sanitize_ndim(subarr, data, dtype, index, allow_2d=allow_2d)\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/pandas/core/construction.py\", line 718, in _sanitize_ndim\n",
      "    raise ValueError(\n",
      "ValueError: Data must be 1-dimensional, got ndarray of shape (2462, 1) instead\n",
      "::1 - - [2025-04-09 17:38:48] \"POST /_dash-update-component HTTP/1.1\" 500 426 0.030168\n",
      "[*********************100%***********************]  1 of 1 completed\n",
      "[2025-04-09 17:38:56,161] ERROR in app: Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/flask/app.py\", line 2077, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/flask/app.py\", line 1525, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/flask/app.py\", line 1523, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/flask/app.py\", line 1509, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/dash/dash.py\", line 1336, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/dash/_callback.py\", line 151, in add_context\n",
      "    output_value = func(*func_args, **func_kwargs)  # %% callback invoked %%\n",
      "                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/workspaces/CS329E/stock-prediction-dash/app.py\", line 150, in stock_price\n",
      "    fig = get_stock_price_fig(df)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/workspaces/CS329E/stock-prediction-dash/app.py\", line 19, in get_stock_price_fig\n",
      "    fig = px.line(df,\n",
      "          ^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/plotly/express/_chart_types.py\", line 260, in line\n",
      "    return make_figure(args=locals(), constructor=go.Scatter)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/plotly/express/_core.py\", line 1933, in make_figure\n",
      "    args = build_dataframe(args, constructor)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/plotly/express/_core.py\", line 1405, in build_dataframe\n",
      "    df_output, wide_id_vars = process_args_into_dataframe(\n",
      "                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/plotly/express/_core.py\", line 1222, in process_args_into_dataframe\n",
      "    df_output[col_name] = to_unindexed_series(df_input[argument])\n",
      "                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/plotly/express/_core.py\", line 1073, in to_unindexed_series\n",
      "    return pd.Series(x).reset_index(drop=True)\n",
      "           ^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/pandas/core/series.py\", line 584, in __init__\n",
      "    data = sanitize_array(data, index, dtype, copy)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/pandas/core/construction.py\", line 633, in sanitize_array\n",
      "    return sanitize_array(\n",
      "           ^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/pandas/core/construction.py\", line 659, in sanitize_array\n",
      "    subarr = _sanitize_ndim(subarr, data, dtype, index, allow_2d=allow_2d)\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/pandas/core/construction.py\", line 718, in _sanitize_ndim\n",
      "    raise ValueError(\n",
      "ValueError: Data must be 1-dimensional, got ndarray of shape (2462, 1) instead\n",
      "::1 - - [2025-04-09 17:38:56] \"POST /_dash-update-component HTTP/1.1\" 500 426 0.029133\n",
      "[*********************100%***********************]  1 of 1 completed\n",
      "[2025-04-09 17:39:10,435] ERROR in app: Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/flask/app.py\", line 2077, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/flask/app.py\", line 1525, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/flask/app.py\", line 1523, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/flask/app.py\", line 1509, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/dash/dash.py\", line 1336, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/dash/_callback.py\", line 151, in add_context\n",
      "    output_value = func(*func_args, **func_kwargs)  # %% callback invoked %%\n",
      "                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/workspaces/CS329E/stock-prediction-dash/app.py\", line 150, in stock_price\n",
      "    fig = get_stock_price_fig(df)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/workspaces/CS329E/stock-prediction-dash/app.py\", line 19, in get_stock_price_fig\n",
      "    fig = px.line(df,\n",
      "          ^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/plotly/express/_chart_types.py\", line 260, in line\n",
      "    return make_figure(args=locals(), constructor=go.Scatter)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/plotly/express/_core.py\", line 1933, in make_figure\n",
      "    args = build_dataframe(args, constructor)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/plotly/express/_core.py\", line 1405, in build_dataframe\n",
      "    df_output, wide_id_vars = process_args_into_dataframe(\n",
      "                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/plotly/express/_core.py\", line 1222, in process_args_into_dataframe\n",
      "    df_output[col_name] = to_unindexed_series(df_input[argument])\n",
      "                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/plotly/express/_core.py\", line 1073, in to_unindexed_series\n",
      "    return pd.Series(x).reset_index(drop=True)\n",
      "           ^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/pandas/core/series.py\", line 584, in __init__\n",
      "    data = sanitize_array(data, index, dtype, copy)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/pandas/core/construction.py\", line 633, in sanitize_array\n",
      "    return sanitize_array(\n",
      "           ^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/pandas/core/construction.py\", line 659, in sanitize_array\n",
      "    subarr = _sanitize_ndim(subarr, data, dtype, index, allow_2d=allow_2d)\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/pandas/core/construction.py\", line 718, in _sanitize_ndim\n",
      "    raise ValueError(\n",
      "ValueError: Data must be 1-dimensional, got ndarray of shape (2462, 1) instead\n",
      "::1 - - [2025-04-09 17:39:10] \"POST /_dash-update-component HTTP/1.1\" 500 426 0.033389\n",
      "[*********************100%***********************]  1 of 1 completed\n",
      "[2025-04-09 17:39:10,575] ERROR in app: Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/flask/app.py\", line 2077, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/flask/app.py\", line 1525, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/flask/app.py\", line 1523, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/flask/app.py\", line 1509, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/dash/dash.py\", line 1336, in dispatch\n",
      "    response.set_data(func(*args, outputs_list=outputs_list))\n",
      "                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/dash/_callback.py\", line 151, in add_context\n",
      "    output_value = func(*func_args, **func_kwargs)  # %% callback invoked %%\n",
      "                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/workspaces/CS329E/stock-prediction-dash/app.py\", line 150, in stock_price\n",
      "    fig = get_stock_price_fig(df)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/workspaces/CS329E/stock-prediction-dash/app.py\", line 19, in get_stock_price_fig\n",
      "    fig = px.line(df,\n",
      "          ^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/plotly/express/_chart_types.py\", line 260, in line\n",
      "    return make_figure(args=locals(), constructor=go.Scatter)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/plotly/express/_core.py\", line 1933, in make_figure\n",
      "    args = build_dataframe(args, constructor)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/plotly/express/_core.py\", line 1405, in build_dataframe\n",
      "    df_output, wide_id_vars = process_args_into_dataframe(\n",
      "                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/plotly/express/_core.py\", line 1222, in process_args_into_dataframe\n",
      "    df_output[col_name] = to_unindexed_series(df_input[argument])\n",
      "                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/plotly/express/_core.py\", line 1073, in to_unindexed_series\n",
      "    return pd.Series(x).reset_index(drop=True)\n",
      "           ^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/pandas/core/series.py\", line 584, in __init__\n",
      "    data = sanitize_array(data, index, dtype, copy)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/pandas/core/construction.py\", line 633, in sanitize_array\n",
      "    return sanitize_array(\n",
      "           ^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/pandas/core/construction.py\", line 659, in sanitize_array\n",
      "    subarr = _sanitize_ndim(subarr, data, dtype, index, allow_2d=allow_2d)\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/codespace/.python/current/lib/python3.12/site-packages/pandas/core/construction.py\", line 718, in _sanitize_ndim\n",
      "    raise ValueError(\n",
      "ValueError: Data must be 1-dimensional, got ndarray of shape (2462, 1) instead\n",
      "::1 - - [2025-04-09 17:39:10] \"POST /_dash-update-component HTTP/1.1\" 500 426 0.032057\n",
      "::1 - - [2025-04-09 17:39:13] \"POST /_dash-update-component HTTP/1.1\" 200 1892 0.144843\n",
      "::1 - - [2025-04-09 17:39:13] \"POST /_dash-update-component HTTP/1.1\" 200 170 0.000457\n",
      "::1 - - [2025-04-09 17:39:13] \"POST /_dash-update-component HTTP/1.1\" 200 166 0.000406\n",
      "::1 - - [2025-04-09 17:39:13] \"POST /_dash-update-component HTTP/1.1\" 200 168 0.000425\n",
      "[*********************100%***********************]  1 of 1 completed\n",
      "/home/codespace/.python/current/lib/python3.12/site-packages/_plotly_utils/basevalidators.py:107: FutureWarning:\n",
      "\n",
      "The behavior of DatetimeProperties.to_pydatetime is deprecated, in a future version this will return a Series containing python datetime objects instead of an ndarray. To retain the old behavior, call `np.array` on the result\n",
      "\n",
      "::1 - - [2025-04-09 17:39:14] \"POST /_dash-update-component HTTP/1.1\" 200 107647 0.270333\n",
      "::1 - - [2025-04-09 17:39:15] \"GET /_dash-component-suites/dash/dcc/async-graph.js HTTP/1.1\" 304 66 0.000529\n",
      "::1 - - [2025-04-09 17:39:15] \"GET /_dash-component-suites/dash/dcc/async-plotlyjs.js HTTP/1.1\" 304 66 0.003692\n",
      "/workspaces/CS329E/stock-prediction-dash/model.py:3: UserWarning:\n",
      "\n",
      "\n",
      "The dash_core_components package is deprecated. Please replace\n",
      "`import dash_core_components as dcc` with `from dash import dcc`\n",
      "\n",
      "/workspaces/CS329E/stock-prediction-dash/model.py:4: UserWarning:\n",
      "\n",
      "\n",
      "The dash_html_components package is deprecated. Please replace\n",
      "`import dash_html_components as html` with `from dash import html`\n",
      "\n",
      "[*********************100%***********************]  1 of 1 completed\n",
      "::1 - - [2025-04-09 17:39:46] \"POST /_dash-update-component HTTP/1.1\" 200 8006 6.504976\n",
      "[*********************100%***********************]  1 of 1 completed\n",
      "::1 - - [2025-04-09 17:39:51] \"POST /_dash-update-component HTTP/1.1\" 200 8006 5.147886\n",
      "[*********************100%***********************]  1 of 1 completed\n",
      "::1 - - [2025-04-09 17:39:57] \"POST /_dash-update-component HTTP/1.1\" 200 8006 5.385973\n"
     ]
    }
   ],
   "source": [
    "! python3 app.py\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
